pkg/kubelet/logs/container_log_manager.go

```diff
@@ -213,80 +223,121 @@ func (c *containerLogManager) Clean(ctx context.Context, containerID string) err
 	for _, container := range containers {
 		// Only rotate logs for running containers. Non-running containers won't
 		// generate new output, it doesn't make sense to keep an empty latest log.
 		if container.GetState() != runtimeapi.ContainerState_CONTAINER_RUNNING {
 			continue
 		}
-		id := container.GetId()
-		// Note that we should not block log rotate for an error of a single container.
-		resp, err := c.runtimeService.ContainerStatus(ctx, id, false)
-		if err != nil {
-			klog.ErrorS(err, "Failed to get container status", "containerID", id)
-			continue
-		}
-		if resp.GetStatus() == nil {
-			klog.ErrorS(err, "Container status is nil", "containerID", id)
-			continue
+		// Doing this to avoid additional overhead with logging of label like arguments that can prove costly
+		if v := klog.V(4); v.Enabled() {
+			klog.V(4).InfoS("Adding new entry to the queue for processing", "id", container.GetId(), "name", container.Metadata.GetName(), "labels", container.GetLabels())
 		}
-		path := resp.GetStatus().GetLogPath()
-		info, err := c.osInterface.Stat(path)
-		if err != nil {
-			if !os.IsNotExist(err) {
-				klog.ErrorS(err, "Failed to stat container log", "path", path)
-				continue
-			}
-			// In rotateLatestLog, there are several cases that we may
-			// lose original container log after ReopenContainerLog fails.
-			// We try to recover it by reopening container log.
-			if err := c.runtimeService.ReopenContainerLog(ctx, id); err != nil {
-				klog.ErrorS(err, "Container log doesn't exist, reopen container log failed", "containerID", id, "path", path)
-				continue
-			}
-			// The container log should be recovered.
-			info, err = c.osInterface.Stat(path)
-			if err != nil {
-				klog.ErrorS(err, "Failed to stat container log after reopen", "path", path)
-				continue
-			}
+		c.queue.Add(container.GetId())
+	}
+	wait.PollInfinite(1*time.Second, func() (done bool, err error) {
+		return c.queue.Len() == 0, nil
+	})
+	klog.V(4).InfoS("Finished container log rotation sequence")
+	return nil
+}
+
+func (c *containerLogManager) processContainer(ctx context.Context, worker int) (ok bool) {
+	// Only rotate logs for running containers. Non-running containers won't
+	// generate new output, it doesn't make sense to keep an empty latest log.
+	// Note that we should not block log rotate for an error of a single container.
+	// In rotateLatestLog, there are several cases that we may
+	// lose original container log after ReopenContainerLog fails.
+	// We try to recover it by reopening container log.
+	// The container log should be recovered.
+	// Perform log rotation.
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer func() {
+		c.queue.Done(key)
+		c.queue.Forget(key)
+	}()
+	// Always default the return to true to keep the processing of Queue ongoing
+	ok = true
+	id := key.(string)
+
+	resp, err := c.runtimeService.ContainerStatus(ctx, id, false)
```

Comment:
Looks like everything here and below is unchanged other than additional logging. A couple comments on the logging.

Not sure how common the error cases are, but if there's any reason those cases could be hit regularly, could blow up the number of logs, but seems like useful information.

Reply:
Agreed. Some of these logs were for me to test some corner cases. I was trying to reproduce some error cases with rotation by making files immutable and such. I am fine cleaning up these logs.